### DISCUSSION TITLE
MQTT5 in Lambda, multi-initialization issue

### DISCUSSION THREAD

**Discussion by SlavenIvanov** [aws/aws-iot-device-sdk-js-v2#360] [Q&A]
Hello,

We are trying to use `aws-iot-device-sdk-js-v2` library in AWS Lambda, because of missing features in the [v3 js sdk](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/) (mainly subscribing to an mqtt topic).

We have created a file that makes a singleton out of an mqtt5 client:

```js
const cfg = {
  endpoint: '213421342134-ats.iot.2341234.amazonaws.com',
  key: fs.readFileSync(path.resolve(__dirname, 'slaven-osx-thing-private.pem.key'), 'utf8'),
  cert: fs.readFileSync(path.resolve(__dirname, 'slaven-osx-thing-certificate.pem.crt'), 'utf8')
}

const builder = iot.AwsIotMqtt5ClientConfigBuilder.newDirectMqttBuilderWithMtlsFromMemory(cfg.endpoint, cfg.cert, cfg.key)
const config: mqtt5.Mqtt5ClientConfig = builder.build()

export const client = new mqtt5.Mqtt5Client(config)
```

And then we import it into a Lambda handler:
```js
export const handler = async () => {
  const connectionSuccess = once(client, 'connectionSuccess')
  client.start()
  await connectionSuccess
  console.log('🟢Connected to AWS IoT')

  const stopped = once(client, 'stopped')
  client.stop()
  await stopped
  console.log('Stopped client')

  client.close()
  console.log('Closed client')

   ....
}
```

The first time we call the Lambda everything goes as expected. The client connects, then disconnects
The second time we call it we get this:
```
"Error: Aws-crt-nodejs does not yet support multi-initialization.",
        "at Object.Module._extensions..node (node:internal/modules/cjs/loader:1189:18)",
        "at Module.load (node:internal/modules/cjs/loader:981:32)",
        "at Function.Module._load (node:internal/modules/cjs/loader:822:12)",
        "at Module.require (node:internal/modules/cjs/loader:1005:19)",
        "at require (node:internal/modules/cjs/helpers:102:18)",
        "at Object.<anonymous> (/Users/slavi/git/growy/growy-v3-lcm/node_modules/aws-crt/dist/native/binding.js:59:19)",
        "at Module._compile (node:internal/modules/cjs/loader:1105:14)",
        "at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)",
        "at Module.load (node:internal/modules/cjs/loader:981:32)"
```

Does anyone have any ideas on a workaround?

**Comment by bretambrose** [comment]
On possibility would be to try and webpack the lambda which should switch you to the browser implementation which does not use a NAPI native add on.

**Reply by SlavenIvanov** [reply]
I am using webpack with serverless. And even if I switch the `target` from `node` to `web` in webpack.config.js , the error persists

**Comment by bretambrose** [comment]
On second look, it seems like the distribution switch gets done in the typescript compilation step:

https://github.com/awslabs/aws-crt-nodejs/blob/main/samples/browser/http/tsconfig.json#L40

If you're using typescript then a copy-paste of that module mapping could be a possibility.  If you're not, there may be some pre-processing tool like babel that could give you a similar transformation on pure javascript.

I don't think it would help, but does anything change if the client is internal to the function?  Once you close the client, it's not usable anymore and with it being a global that feels a little weird.  If you didn't call close at the end of the function, in theory the cached client may still be usable on a subsequent lambda invocation by a new call to start.

When run in node, the SDK uses a NAPI addon to pull in native implementations of a lot of functionality (mqtt, http, auth/crypto, etc...).  That native addon does not support being initialized more than once, which is the core problem you're running into.  In the past, we only saw that behavior when used from a framework (Electron and similar could have it happen) that potentially ran more than one node instance in a single process.  I don't know what Lambda does to try and avoid paying the cold start cost of launching node, but it seems unlikely that they'd spawn a second instance in the same sandboxed process.

I'm sick right now, but when I'm coherent and able, I can also open an internal ticket to Lambda to find out more about how they cache across invocations within the same sandbox process.  That might give us some clues as to an alternate workaround.

The long-term fix of allowing multi-init of the NAPI module is potentially a substantial project and I would be concerned of use cases that triggered it arbitrarily like this because each module init of the CRT spawns multiple long-lived threads.

Another possibility is that the multi-init problem you're facing isn't actually concurrent, instead when the module is finalized we forget to reset it to allow another init to happen (Lambda keeps node running).  Checking that...

Yeah, there is no reset at the moment.  Comparing the multi-init check (https://github.com/awslabs/aws-crt-nodejs/blob/main/source/module.c#L1139-L1159) with the finalizer (https://github.com/awslabs/aws-crt-nodejs/blob/main/source/module.c#L1077-L1102), it's clear we don't do any kind of reset on the possibility of a re-initialization.  That might actually work for the Lambda case, and would be a lot less work than the concurrent multi-init story.  I can look into it.

**Reply by SlavenIvanov** [reply]
Thank you for taking the time, and get well soon!

I tried the proposed solutions, so I wanted to give you an update on those:

1. Does anything change if the client is internal to the function?
- Didn't change anything
2. Using the browser implementation.
I tried using:
```js
import { iot, mqtt5 } from 'aws-iot-device-sdk-v2/dist/browser'
const builder = iot.AwsIotMqtt5ClientConfigBuilder.newWebsocketMqttBuilderWithCustomAuth( .... )
const clientInstance: mqtt5.Mqtt5Client = new mqtt5.Mqtt5Client(builder.build())
```
With this the first time we call the lambda we get:
```
✖ ReferenceError: window is not defined
    at Object.<anonymous> (.../.webpack/service/src/functions/api/TestMqttAuth.js:6032:1)
    at __webpack_require__ (.../.webpack/service/src/functions/api/TestMqttAuth.js:45208:42)
    at Object.<anonymous> (.../.webpack/service/src/functions/api/TestMqttAuth.js:5986:1)
    at __webpack_require__ (.../.webpack/service/src/functions/api/TestMqttAuth.js:45208:42)
    at Object.<anonymous> (.../.webpack/service/src/functions/api/TestMqttAuth.js:2687:19)
    at __webpack_require__ (.../.webpack/service/src/functions/api/TestMqttAuth.js:45208:42)
    at Module.Object.create.k2 (.../.webpack/service/src/functions/api/TestMqttAuth.js:2611:92)
    at __webpack_require__ (.../.webpack/service/src/functions/api/TestMqttAuth.js:45208:42)
    at .../.webpack/service/src/functions/api/TestMqttAuth.js:45291:81
    at .../.webpack/service/src/functions/api/TestMqttAuth.js:45319:3
✖ window is not defined
```
And this is the piece of code which errors out:
<img width="680" alt="Screenshot 2023-04-27 at 11 21 40" src="https://user-images.githubusercontent.com/104974346/234803371-44bdd688-d70f-4f43-b087-b6f4bdcaf4ec.png">
This persists even with `target: 'web',` in the webpack config.
And then on the second handler invocation, the "multi-initialization" error is thrown.

**Comment by bretambrose** [comment]
How are you bundling the SDK into your lambda?  I'm trying to repro, but running into the issue of getting the napi module to be found (AWS CRT binary not present in /bin/linux-x64/aws-crt-nodejs).  In the interests of time-saving, following what you're doing may make repro quicker.

On a more general note, we need better support in https://github.com/awslabs/aws-crt-nodejs/blob/main/lib/native/binding.js for alternative load paths for the NAPI modules.  Just reading the code you can see how it will fail based on the directory assumptions once something has been packed into a unity javascript file.  Unsure what the right way forward there is.

**Reply by SlavenIvanov** [reply]
Hi, I am working on a reproduction repo, and will post it here when done.

**Reply by SlavenIvanov** [reply]
Hello again
We haven't reached the part where we deploy it on AWS, all of the tests so far have been done locally so far.

I've made a minimal repro repo for our setup (no typescript, webpack, just serverless): https://github.com/SlavenGrowy/lambda-iot-device-sdk-repro

**Comment by bretambrose** [comment]
Thank your for the extremely easy repro setup.  I have it on my local system and can now experiment with potential solutions

**Comment by bretambrose** [comment]
So good news and bad news:

The bad - this is a genuine case of multi-init (not a reload of an unloaded module like I had guessed above).  This won't be an easy fix.  I know in theory how to do it, but the worry is that it will introduce some instability once we no longer have singletons for key resources.  It won't be a huge delta, but it will be a delta in very delicate parts of the codebase.

The good - this is such an easy repro case and this is such an unpleasant surprise, that I'm going to try and scope out the work to do the full multi-init fix.  

That being said, even if we address multi-init successfully, I have some significant concerns about the overall approach given that each invocation appears to be doing a full module load of the CRT.  Underneath the hood, that means at least two threads spun up (one in the event loop group, one (or more) for DNS resolution) and those do not go away until node finalizes the module which we have no control over and may just not happen.  Depending on the concurrency of lambda invocation, this could be a little ugly when scaled up.

**Comment by bretambrose** [comment]
An update:

I have a proposed fix for multi-init here: https://github.com/awslabs/aws-crt-nodejs/pull/451

The fix attempts to stay minimal and simple.  In particular, no additional threads are going to spin up (other than the node worker thread) when multi-init occurs.  All node worker threads (+ the main thread) will shared a single set of CRT resources and only when they all drop their references will a complete unload/cleanup occur.  So some of my concerns in the reply above shouldn't be an issue.

However, I've spent most of the last few days tracking down a related issue that I'm guessing you haven't encountered yet.

Your repro, on a glibc-based Linux host (musl is unaffected), reliably crashes when the last worker thread that has a reference to the NodeJS CRT terminates.  The reason for this is that some dependencies of the CRT have a bug (s2n) or flat out don't support (aws-lc) proper cleanup of the thread locals that they use.

For s2n, you can see the related issues here:
https://github.com/aws/s2n-tls/issues/3987
https://github.com/aws/s2n-tls/issues/3990

For aws-lc, without a public API for dealing with init/cleanup of thread-local data, we may end up having to move off of it as a dependency and instead use openssl's libcrypto (which has proper init/cleanup and doesn't crash).

The upshot is that even with the multi-init fix, if you're running serverless and the JS CRT on a glibc-based Linux distribution, you will crash.  My guess is you're probably developing on Mac or Windows and that's why you haven't complained about this.

First steps first though, I'll try and get the multi-init PR pushed through to the SDK.

**Reply by SlavenIvanov** [reply]
Hi Thanks for the update! You are right. We are mostly using OSX. But we would have found that later, when we would try to get it running in the cloud.

**Comment by bretambrose** [comment]
Related/Same Issue (worker shut down crash in linux): https://github.com/awslabs/aws-crt-nodejs/issues/452

**Comment by bretambrose** [comment]
As an update, the latest SDK release should fix the multi-init issue.  We are still working with our dependencies to address the linux/glibc crash.

**Comment by bretambrose** [comment]
I've tested the proposed fixes from aws-lc and s2n and have verified that they resolve the crash.  There's a separate resource leak (thread local key) issue that is also under consideration but it's slow and less pressing than resolving the crash.

**Comment by bretambrose** [comment]
The crash on unload should be fixed in https://github.com/aws/aws-iot-device-sdk-js-v2/releases/tag/v1.13.4

**Reply by jmklix** [reply]
@SlavenIvanov are you running into any more problems after using the above fix?

**Reply by SlavenIvanov** [reply]
Hello @bretambrose @jmklix , thanks for the update.I am currently working on a separate task, after which I will be able to test this out. I will give you an update as soon as I try it out.

**Comment by SlavenIvanov** [comment]
Hi @bretambrose , @jmklix I had the time to test out the fixes, here is what I found:

**Local testing consisting of:**
- One Lambda that does:
```typescript
  const connectionSuccess = once(client, 'connectionSuccess')
  client.start()
  await connectionSuccess
  console.log('🟢Connected to AWS IoT')
  await client.publish({...})
  console.log('🟢Published message👌')

  const stopped = once(client, 'stopped')
  client.stop()
  await stopped
  console.log('Stopped client')

  client.close()
  console.log('Closed client')
```
- Lambda built and packaged by webpack with the following config:
```js
const slsw = require('serverless-webpack')
const path = require('path')

const dev = process.env.NODE_ENV !== 'production'

module.exports = {
  node: {
    __dirname: true
  },
  target: 'node',
  mode: dev ? 'none' : 'production',
  entry: slsw.lib.entries,
  devtool: 'source-map',
  output: {
    libraryTarget: 'commonjs',
    path: path.resolve(__dirname, '.webpack'),
    filename: '[name].js'
  },
  resolve: {
    extensions: ['.ts', '.js']
  },
  module: {
    rules: [{ test: /\.(ts|js)x?$/, exclude: /node_modules/, use: ['babel-loader'] }]
  },
  stats: {
    preset: 'minimal'
  },
  externals: ['aws-sdk', '@aws-sdk/client-dynamodb', '@aws-sdk/lib-dynamodb']
}

```
And webpack versions:
```
    "serverless-webpack": "^5.11.0",
    "webpack": "^5.65.0",
```

1. Node app no longer crashes when Lambda is warm, that is great!
2. Can subscribe and publish to topics sucessfully!
3. When you leave the node app alive for some time, after you have called the iot lambda(even once) that initializes the aws-iot-client-sdk, the app crashes with the following error code:
<img width="816" alt="Screenshot 2023-07-12 at 11 22 46" src="https://github.com/aws/aws-iot-device-sdk-js-v2/assets/104974346/751d223d-8589-49f1-97ff-b62cc43434f9">

That is not good, since it may potentially kill lambdas while they are doing work.

4. App execution fails sporadically with:
 ```
GET /req-res (λ: TestMqttAuth)
✖ Unhandled exception in handler 'TestMqttAuth'.
✖ Runtime.ImportModuleError: Error: Cannot find module '../node_modules/aws-crt/dist/bin/darwin-arm64/aws-crt-nodejs'
      at _loadUserApp (../node_modules/serverless-offline/src/lambda/handler-runner/in-process-runner/aws-lambda-ric/UserFunction.js:310:15)
      at async module.exports.load (../node_modules/serverless-offline/src/lambda/handler-runner/in-process-runner/aws-lambda-ric/UserFunction.js:341:21)
      at async InProcessRunner.run (../node_modules/serverless-offline/src/lambda/handler-runner/in-process-runner/InProcessRunner.js:41:21)
      at async MessagePort.<anonymous> (../node_modules/serverless-offline/src/lambda/handler-runner/worker-thread-runner/workerThreadHelper.js:24:14)
✖ Error: Cannot find module '../node_modules/aws-crt/dist/bin/darwin-arm64/aws-crt-nodejs'
 ```

even with a manual installation of  `"aws-crt": "^1.15.20",`

**Cloud testing consisting of:**
- building and deploying the same app to the AWS cloud

**Cloud results:**
- Whenever the app gets called it fails with:
```
{
  "errorType": "Error",
  "errorMessage": "AWS CRT binary not present in any of the following locations:\n\t/var/task/node_modules/aws-crt/bin/linux-x64/aws-crt-nodejs",
  "trace": [
    "Error: AWS CRT binary not present in any of the following locations:",
    "\t/var/task/node_modules/aws-crt/bin/linux-x64/aws-crt-nodejs",
    "    at Object.<anonymous> (/var/task/src/functions/api/TestMqttAuth.js:1517:11)",
    "    at __webpack_require__ (/var/task/src/functions/api/TestMqttAuth.js:30124:42)",
    "    at Object.<anonymous> (/var/task/src/functions/api/TestMqttAuth.js:1244:35)",
    "    at __webpack_require__ (/var/task/src/functions/api/TestMqttAuth.js:30124:42)",
    "    at Object.<anonymous> (/var/task/src/functions/api/TestMqttAuth.js:854:27)",
    "    at __webpack_require__ (/var/task/src/functions/api/TestMqttAuth.js:30124:42)",
    "    at Object.<anonymous> (/var/task/src/functions/api/TestMqttAuth.js:698:19)",
    "    at __webpack_require__ (/var/task/src/functions/api/TestMqttAuth.js:30124:42)",
    "    at Object.<anonymous> (/var/task/src/functions/api/TestMqttAuth.js:170:50)",
    "    at __webpack_require__ (/var/task/src/functions/api/TestMqttAuth.js:30124:42)"
  ]
}
```
Maybe webpack is stripping the `aws-crt` binaries from the build?

**Comment by bretambrose** [comment]
I'll be starting oncall tomorrow and take a look at these as soon as I can, pending other priorities.

**Comment by bretambrose** [comment]
(@SlavenIvanov) Can you update your repro repo (https://github.com/SlavenIvanov/lambda-iot-device-sdk-repro) with the webpack changes that reproduce the first crash?  I will attempt to apply the deltas you've listed above, but in case something is missing, an authoritative snapshot would be fantastic.

I don't have any insight for the sporadic fail-to-load in serverless.

For webpack + Lambda deploy, it's a known issue that the CRT attempts to load the native module at a hard-coded offset from the loading js file (https://github.com/awslabs/aws-crt-nodejs/blob/main/lib/native/binding.js#L22-L43)  Bundling tools like webpack and others break that expectation.  I'm not sure what the best possible solution is, but here are a couple of workarounds:

1. If you update your webpack config to use the browser implementations, your problems will disappear because the napi module won't be involved anymore.  I'm not sure the exact way to do this via webpack because we perform the browser/node switch in typescript config (https://github.com/aws/aws-iot-device-sdk-js-v2/blob/main/samples/browser/pub_sub_mqtt5/tsconfig.json#L40)
2. Barring that, we could add support for using environment variable(s) to help locate the napi module.  You could bundle/webpack your javascript, and include the appropriate (platform/architecture) napi binary from the unpacked CRT wherever you want to locate it.  Then modify the lambda environment to have the right path offset for the napi module loading to succeed.

So as an example, if you bundle into a single index.js file and included the right binary at the same level:

lambda.zip
|--- index.js
|--- aws-crt-nodejs

Then setting whatever environment variable we settle on to `.` would let the loader logic resolve to the correct location of the napi module

**Comment by SlavenIvanov** [comment]
Hi @bretambrose 

Thank you for taking the time to take a look at this.

First off it was my mistake when I said that the missing binary error is sporadic. The I had falsely assumed that since I had been trying the reproduction in several repositories, one of which was not using webpack to build the project. When I use webpack, the issue is persistant.

Second: I updated the repro: https://github.com/SlavenIvanov/lambda-iot-device-sdk-repro . If you have any issues with getting it running, please let me know.

I have tried to switching the library implementation to the web versions.
Switching is a single property change in `webpack.config.js` : 
```
module.exports = {
....
  target: "web",
.....
};
```
And I get this error:
```
GET /dev/iot (λ: IotFunction)
✖ Unhandled exception in handler 'IotFunction'.
✖ ReferenceError: window is not defined
at Object.<anonymous> (.webpack/service/src/IotFunction.js:835:1)
at __webpack_require__ (.webpack/service/src/IotFunction.js:64530:42)
at Object.<anonymous> (.webpack/service/src/IotFunction.js:789:1)
at __webpack_require__ (.webpack/service/src/IotFunction.js:64530:42)
at Object.<anonymous> (.webpack/service/src/IotFunction.js:643:19)
at __webpack_require__ (.webpack/service/src/IotFunction.js:64530:42)
at Object.<anonymous> (.webpack/service/src/IotFunction.js:115:50)
at __webpack_require__ (.webpack/service/src/IotFunction.js:64530:42)
at Object.<anonymous> (.webpack/service/src/IotFunction.js:44:34)
at __webpack_require__ (.webpack/service/src/IotFunction.js:64530:42)
✖ window is not defined
```
I've also tried defining `var window = {}` at the top of the Lambda, before all imports, to see if it makes a difference, but it doesn't.
Honestly if we can get this easily fixed by using the browser implementation that would be fantastic. I don't know if this solution won't cause future problems with node-only packages we might be using.

Another solution that I tried was to use https://www.npmjs.com/package/node-loader with the following configuration in `webpack.config.js` to bundle the binaries with the project build files:
```
.....
module.exports = {
.....
  module: {
    rules: [
      { test: /\.(ts|js)x?$/, exclude: /node_modules/, use: ["ts-loader"] },
        {
          test: /\.node$/,
          loader: 'node-loader',
          options: {
            name: '[path][name].[ext]'
          }
        }
    ],
  },
....
};
```
And when I build and run the lambdas I still get the missing binaries error. When I sun `serverless package` and inspect the generated. zip file in `.serverless` folder, i can see that the binary is not included.

So I tried to then import the binaries in the lambda as a file, so that It can get recognized by webpack and included into the build, like so:

```
import file from '../../node_modules/aws-crt/dist/bin/darwin-arm64/aws-crt-nodejs.node'

export const handler = async () => {
   console.log(file)
....
}
```
After building I can see that the binaries then get included into the build
<img width="1454" alt="Screenshot 2023-07-19 at 14 25 25" src="https://github.com/aws/aws-iot-device-sdk-js-v2/assets/104974346/b4bbf124-ff4f-4864-9599-88ab7a2c2cd8">
, but still the error persists even though the files exist in the directory that its looking at

```
GET /dev/iot (λ: IotFunction)
✖ Unhandled exception in handler 'IotFunction'.
✖ Runtime.ImportModuleError: Error: Cannot find module '/lambda-iot-device-sdk/node_modules/aws-crt/dist/bin/darwin-arm64/aws-crt-nodejs'
at _loadUserApp (/lambda-iot-device-sdk/node_modules/serverless-offline/src/lambda/handler-runner/in-process-runner/aws-lambda-ric/UserFunction.js:310:15)
at async module.exports.load (/lambda-iot-device-sdk/node_modules/serverless-offline/src/lambda/handler-runner/in-process-runner/aws-lambda-ric/UserFunction.js:341:21)
at async InProcessRunner.run (file:///lambda-iot-device-sdk/node_modules/serverless-offline/src/lambda/handler-runner/in-process-runner/InProcessRunner.js:41:21)
at async MessagePort.<anonymous> (file:///lambda-iot-device-sdk/node_modules/serverless-offline/src/lambda/handler-runner/worker-thread-runner/workerThreadHelper.js:24:14)
✖ Error: Cannot find module '/lambda-iot-device-sdk/node_modules/aws-crt/dist/bin/darwin-arm64/aws-crt-nodejs'
```

Again, thank you for your time.

Best Regards,
Slaven

**Comment by bretambrose** [comment]
Thanks for the update.  Current state of your repo successfully repros the module loading issue but priority-wise I'm much more interested in the napi_release_threadsafe_function crash.  What does it take to make that happen?
